common/src/test/java/com/thoughtworks/go/remote/SerializationTest.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2020 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.remote;
+
+import com.thoughtworks.go.security.*;
+import com.thoughtworks.go.util.SystemEnvironment;
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+
+import static java.lang.String.format;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.mock;
+
+@SuppressWarnings("deprecation")
+class SerializationTest {
+    @Test
+    void rejectsSerializationOfGoCipher() {
+        final IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () ->
+                Serialization.instance().toJson(new GoCipher(mock(Encrypter.class))));
```

Comment:
What happens when we try to serialize `ArtifactStore`, the store needs to be serialized but deep down it does have a cipher.

Reply:
It depends on the `ArtifactStore` type adapter. If it attempts to do a `toJson()` on it rather than computing a set of decrypted secrets, it will blow up.

Reply:
It looks like it would work, unless I'm missing something.

For serialization, we never descend that far into configuration properties. We do this instead:

https://github.com/marques-work/gocd/blob/ff49b4ff9ce81f2379e2382ad8db362ab7e331bc/common/src/main/java/com/thoughtworks/go/remote/Serialization.java#L184-L190

And for deserialization, we do the reverse. The `Gson` instance never encounters a `GoCipher` instance:

https://github.com/marques-work/gocd/blob/ff49b4ff9ce81f2379e2382ad8db362ab7e331bc/common/src/main/java/com/thoughtworks/go/remote/Serialization.java#L197-L199

So we shouldn't blow up here. We can add a test though.